### ISSUE TITLE
ScalaJS linking error when calling addPathPart

### ISSUE THREAD

**Issue by povder** [lemonlabsuk/scala-uri#152]
scala-uri version: 2.2.2
ScalaJS version: 1.0.1
Scala version: 2.13.1

**Describe the bug**
When `addPathPart` and other methods that modify URLs such as "addParams" are used, fastOptJS task returns linking errors:

```
[error] Referring to non-existent class io.lemonlabs.uri.typesafe.PathPart$Ops
[error]   called from io.lemonlabs.uri.typesafe.PathPart$AllOps
[error]   called from io.lemonlabs.uri.typesafe.PathPart$ops$.toAllPathPartOps(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.typesafe.PathPart$AllOps
[error]   called from io.lemonlabs.uri.UrlPath.addPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.UrlPath
[error]   called from io.lemonlabs.uri.Url.addPathPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.Url
[error]   called from io.lemonlabs.Main$.writeUrl(io.lemonlabs.uri.Url)void
[error]   called from io.lemonlabs.Main$.urlTextTyped()void
[error]   called from io.lemonlabs.Main$.main([java.lang.String)void
[error]   called from static io.lemonlabs.Main.main([java.lang.String)void
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   io.lemonlabs.uri.typesafe.PathPart$ops$
[error]   io.lemonlabs.Main$
[error] Referring to non-existent method io.lemonlabs.uri.typesafe.PathPart$AllOps.path()java.lang.String
[error]   called from io.lemonlabs.uri.UrlPath.addPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.UrlPath
[error]   called from io.lemonlabs.uri.Url.addPathPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.Url
[error]   called from io.lemonlabs.Main$.writeUrl(io.lemonlabs.uri.Url)void
[error]   called from io.lemonlabs.Main$.urlTextTyped()void
[error]   called from io.lemonlabs.Main$.main([java.lang.String)void
[error]   called from static io.lemonlabs.Main.main([java.lang.String)void
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   io.lemonlabs.Main$
[error] Referring to non-existent class io.lemonlabs.uri.typesafe.PathPart$Ops
[error]   called from io.lemonlabs.uri.typesafe.PathPart$ops$$anon$2.path()java.lang.String
[error]   called from io.lemonlabs.uri.UrlPath.addPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.UrlPath
[error]   called from io.lemonlabs.uri.Url.addPathPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.Url
[error]   called from io.lemonlabs.Main$.writeUrl(io.lemonlabs.uri.Url)void
[error]   called from io.lemonlabs.Main$.urlTextTyped()void
[error]   called from io.lemonlabs.Main$.main([java.lang.String)void
[error]   called from static io.lemonlabs.Main.main([java.lang.String)void
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   io.lemonlabs.uri.typesafe.PathPart$ops$$anon$2
[error]   io.lemonlabs.Main$
[error] Referring to non-existent method io.lemonlabs.uri.typesafe.PathPart$Ops.path()java.lang.String
[error]   called from io.lemonlabs.uri.typesafe.PathPart$ops$$anon$2.path()java.lang.String
[error]   called from io.lemonlabs.uri.UrlPath.addPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.UrlPath
[error]   called from io.lemonlabs.uri.Url.addPathPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.Url
[error]   called from io.lemonlabs.Main$.writeUrl(io.lemonlabs.uri.Url)void
[error]   called from io.lemonlabs.Main$.urlTextTyped()void
[error]   called from io.lemonlabs.Main$.main([java.lang.String)void
[error]   called from static io.lemonlabs.Main.main([java.lang.String)void
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   io.lemonlabs.uri.typesafe.PathPart$ops$$anon$2
[error]   io.lemonlabs.Main$
[error] Referring to non-existent class io.lemonlabs.uri.typesafe.PathPart$Ops
[error]   called from constructor io.lemonlabs.uri.typesafe.PathPart$ops$$anon$2.<init>(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)void
[error]   called from io.lemonlabs.uri.typesafe.PathPart$ops$.toAllPathPartOps(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.typesafe.PathPart$AllOps
[error]   called from io.lemonlabs.uri.UrlPath.addPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.UrlPath
[error]   called from io.lemonlabs.uri.Url.addPathPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.Url
[error]   called from io.lemonlabs.Main$.writeUrl(io.lemonlabs.uri.Url)void
[error]   called from io.lemonlabs.Main$.urlTextTyped()void
[error]   called from io.lemonlabs.Main$.main([java.lang.String)void
[error]   called from static io.lemonlabs.Main.main([java.lang.String)void
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   io.lemonlabs.uri.typesafe.PathPart$ops$
[error]   io.lemonlabs.Main$
[error] Referring to non-existent method io.lemonlabs.uri.typesafe.PathPart$Ops.$init$()void
[error]   called from constructor io.lemonlabs.uri.typesafe.PathPart$ops$$anon$2.<init>(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)void
[error]   called from io.lemonlabs.uri.typesafe.PathPart$ops$.toAllPathPartOps(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.typesafe.PathPart$AllOps
[error]   called from io.lemonlabs.uri.UrlPath.addPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.UrlPath
[error]   called from io.lemonlabs.uri.Url.addPathPart(java.lang.Object,io.lemonlabs.uri.typesafe.PathPart)io.lemonlabs.uri.Url
[error]   called from io.lemonlabs.Main$.writeUrl(io.lemonlabs.uri.Url)void
[error]   called from io.lemonlabs.Main$.urlTextTyped()void
[error]   called from io.lemonlabs.Main$.main([java.lang.String)void
[error]   called from static io.lemonlabs.Main.main([java.lang.String)void
[error]   called from core module module initializers
[error] involving instantiated classes:
[error]   io.lemonlabs.uri.typesafe.PathPart$ops$
[error]   io.lemonlabs.Main$
[error] There were linking errors
[error] (Compile / fastOptJS) There were linking errors
```

**To Reproduce**
1. git clone https://github.com/lemonlabsuk/scala-uri-scalajs-example.git (commit https://github.com/lemonlabsuk/scala-uri-scalajs-example/commit/3175f5865bb9a3e933acf99d3b8c7b39d59109cd)
1. in `src/main/scala/io/lemonlabs/Main.scala` file after line 62 add `url.addPathPart("part")`.
1. Start sbt and execute `fastOptJS` task.

**Comment by theon** [comment]
Interesting; thanks for raising.

I thought I was the only one, but it appears not. See this comment where this has been previously discussed: https://github.com/lemonlabsuk/scala-uri/issues/117#issuecomment-626018628

For some reason this doesn't happen on travis CI and doesn't happen for everyone. I think there must be some environmental element to this, but I'm not sure what.

**Comment by povder** [comment]
Thanks for the quick reply and for sharing the link. I'll try to find what may be wrong with my environment and if I do find it I'll comment here.

**Comment by povder** [comment]
Interestingly, the scala-uri's tests succeed for me using `SCALAJS_VERSION=1.0.1 sbt clean coverageOff ++2.13.1 scalaUriJS/test` command. I don't have the problem that you mentioned in https://github.com/lemonlabsuk/scala-uri/issues/117#issuecomment-626018628. 

scala-uri lib consumers don't build for me (my own project and the scala-uri demo).

**Comment by theon** [comment]
I'm getting more confused, the more we learn 😆

**Comment by povder** [comment]
I bumped the version of scala-uri locally to 2.2.3-SNAPSHOT and did `++2.13.1 scalaUriJS/publishLocal`. Then I used that locally published version in the `scala-uri-scalajs-example` project and it compiles fine now lol.

I don't know, I'll run a diff between jars later, I still have no idea what the problem is/was.

**Comment by theon** [comment]
I tried inlining all the simulacrum generated code to play with the traits the Scala.js linker is complaining about: https://github.com/lemonlabsuk/scala-uri/commit/a49ced191a6fcd447e49d397db61048e7ce653cb

If I rename all the `trait Ops[A]` to something else, e.g. `trait BaseOps[A]` I am able to run the command below. Renaming back to `Ops` and the Scala.js linking fails again

```
SCALAJS_VERSION=1.1.0 sbt clean coverageOff ++2.13.1 scalaUriJS/test
```

So it seems to be something to do with the name of those `Ops` traits. Still not sure why the linking when running the tests are only failing for me

**Comment by theon** [comment]
Note: I'm not intending to merge that inlined generated code (it will be a nightmare to maintain)... I'm just using it to diagnose the issue

**Comment by takayahilton** [comment]
Maybe we can use [this](https://github.com/typelevel/simulacrum-scalafix)?

**Comment by theon** [comment]
> Maybe we can use [this](https://github.com/typelevel/simulacrum-scalafix)?

Certainly worth a try; I would be happy if it solves the problem.

**Comment by easel** [comment]
For reference, I'm experiencing this issue blocking my port to scala.js 1.0 on scala 2.12 as well.

Unfortunately, at least out of the box, scalafix-simulacrum isn't picking up the annotations. Chances are it's operator error on my part, but if anybody wants to give it a try, take a look at https://github.com/easel/scala-uri/tree/scalafix-simulacrum

**Comment by easel** [comment]
I can confirm that typelevel/simulacrum#202 fixes this issue, and that is exclusive to MacOS (and probably windows). It does not occur on linux.

**Comment by theon** [comment]
Oh awesome, thanks for your work on this  @easel 

I'll upgrade ASAP. I'll also have a play with your migration to the scalafix version as it sounds like the fix is released over there (and I think that would be a good upgrade anyway)

**Comment by theon** [comment]
Thanks again @easel @takayahilton 

This has been released as version `2.2.3` and should be in maven central shortly
